lib/pure/bitops.nim

```diff
@@ -163,174 +164,208 @@ elif useICC_builtins:
+      when sizeof(x) <= 2: builtin_popcnt16(x.uint16).int
+      elif sizeof(x) <= 4: builtin_popcnt32(x.uint32).int
+      elif arch64:         builtin_popcnt64(x.uint64).int
+      else:                builtin_popcnt32((x.uint64 and 0xFFFFFFFF'u64).uint32 ).int +
+                          builtin_popcnt32((x.uint64 shr 32'u64).uint32 ).int
     elif useICC_builtins:
-      when sizeof(x) <= 4: result = builtin_popcnt32(x.cint).int
-      elif arch64:         result = builtin_popcnt64(x.uint64).int
-      else:                result = builtin_popcnt32((x.uint64 and 0xFFFFFFFF'u64).cint ).int +
-                                    builtin_popcnt32((x.uint64 shr 32'u64).cint ).int
+      when sizeof(x) <= 4: builtin_popcnt32(x.cint).int
+      elif arch64:         builtin_popcnt64(x.uint64).int
+      else:                builtin_popcnt32((x.uint64 and 0xFFFFFFFF'u64).cint ).int +
+                          builtin_popcnt32((x.uint64 shr 32'u64).cint ).int
     else:
-      when sizeof(x) <= 4: result = countSetBits_nim(x.uint32)
-      else:                result = countSetBits_nim(x.uint64)
+      when sizeof(x) <= 4: oneBitsNim(x.uint32)
+      else:                oneBitsNim(x.uint64)
+
+func countSetBits*(x: SomeInteger): int {.deprecated: "oneBits".} =
+  oneBits(x.uint64)
 
-proc popcount*(x: SomeInteger): int {.inline, nosideeffect.} =
+func popcount*(x: SomeInteger): int {.deprecated: "oneBits".}=
   ## Alias for for countSetBits (Hamming weight.)
-  result = countSetBits(x)
+  oneBits(x.uint64)
 
-proc parityBits*(x: SomeInteger): int {.inline, nosideeffect.} =
+func parityBits*(x: SomeUnsignedInt): int =
   ## Calculate the bit parity in integer. If number of 1-bit
   ## is odd parity is 1, otherwise 0.
+  ##
+  ## Example:
+  ## doAssert parityBits(0b00000001'u8) == 1
   # Can be used a base if creating ASM version.
   # https://stackoverflow.com/questions/21617970/how-to-check-if-value-has-even-parity-of-bits-or-odd
   when nimvm:
-    when sizeof(x) <= 4: result = parity_impl(x.uint32)
-    else:                result = parity_impl(x.uint64)
+    when sizeof(x) <= 4: parityImpl(x.uint32)
+    else:                parityImpl(x.uint64)
   else:
     when useGCC_builtins:
-      when sizeof(x) <= 4: result = builtin_parity(x.uint32).int
-      else:                result = builtin_parityll(x.uint64).int
+      when sizeof(x) <= 4: builtin_parity(x.uint32).int
+      else:                builtin_parityll(x.uint64).int
     else:
-      when sizeof(x) <= 4: result = parity_impl(x.uint32)
-      else:                result = parity_impl(x.uint64)
+      when sizeof(x) <= 4: parityImpl(x.uint32)
+      else:                parityImpl(x.uint64)
+
+func parityBits*(x: SomeSignedInt): int {.deprecated.} =
+  parityBits(x.uint64)
 
-proc firstSetBit*(x: SomeInteger): int {.inline, nosideeffect.} =
+func firstOneBit*(x: SomeUnsignedInt, maybeZero = true): int =
```

Comment:
This should be a static bool. I don't se any use-case where having it runtime defined is useful.

Reply:
I would keep the inline as well here. In a worse case scenario the operations are:

on x86:
BSR + XOR + SUB (if the compiler doesn't cancel out the sub and xor)

With a non-inline function you would have to add push, pop, ret and doubling the cost with no code size advantage.

Reply:
`static[bool]`: there's no semantic reason for it and it makes the declaration a lot more ugly - requires users of the api to waste more effort on understanding what it means. compilers nowadays do interprocedural constant propagation, ie https://kristerw.blogspot.com/2017/05/interprocedural-optimization-in-gcc.html. honestly, the best I can think of is actually to remove the parameter completely

`inline`: as noted in intro, a feature that adds visual noise and where the compiler is generally in a better spot to make the right tradeoff, based on contextual information (target instruction set, size-vs-speed flags, other operations near the call, constant folding, subexpression eval etc etc.. the only reason to keep it really is that nim doesn't use lto or amalgamation by default meaning that the way the c backend is setup, inlining won't happen across module boundaries which is a shame, also for constant propagation